Scouting/DS Stuff/2014 Scouting/config.txt
Configuration
	InputFile: DS.txt # The input file of match numbers and teams playing.
	Foo: file.out # The file to put the output of the scouting.

Screens
	1:HomeScreen
	2:AutoScreen
	3:TeleScreen
	4:TeleScreen2

HomeScreen
	name: Home
	ShowScoutingInitial

AutoScreen
	name: Auto
	PlainText
		x:0
		y:5
		value:"No Show/Dead Bot"
	Checkbox
		x:16
		y:4
		name:autoDead
	PlainText
		x:0
		y:7
		value:"Low"
	Counter
		x:4
		y:7
		increment:x
		name:autoLow
	PlainText
		x:9
		y:7
		value:"Low Hot"
	Counter
		x:17
		y:7
		increment:y
		name:autoLowH
	PlainText
		x:0
		y:9
		value:"High"
	Counter
		x:4
		y:9
		increment:a
		name:autoHigh
	PlainText
		x:9
		y:9
		value:"High Hot"
	Counter
		x:17
		y:9
		increment:b
		name:autoHighH
	PlainText
		x:0
		y:12
		value:"Mobility"
	Checkbox
		x:8 
		y:11
		name:autoMobility
	PlainText
		x:0
		y:14
		value:"Missed Low"
	Counter
		x:11
		y:14
		increment:L
		name:autoMissedL
	PlainText
		x:0
		y:16
		value:"Missed High"
	Counter
		x:12
		y:16
		increment:R
		name:autoMissedH

TeleScreen
	name: Tele
	PlainText
		x:0
		y:5
		value:"Low"
	Counter
		x:4
		y:5
		increment:x
		name:teleLow
	PlainText
		x:9
		y:5
		value:"High"
	Counter
		x:14
		y:5
		increment:y
		name:teletop
	PlainText
		x:0
		y:7
		value:"2 Assist:"
	Counter
		x:9
		y:7
		increment:a
		name:2assist
	PlainText
		x:0
		y:9
		value:"3 Assist:"
	Counter
		x:9
		y:9
		increment:b
		name:3assist
	PlainText
		x:0
		y:11
		value:"Possession:"
	Rating
		x:11
		y:11
		name:possession
	PlainText
		x:0
		y:13
		value:"Truss"
	Counter
		x:6
		y:13
		increment:L
		value:truss
	PlainText
		x:0
		y:15
		value:"Catch"
	Counter
		x:6
		y:15
		increment:R
		value:catch
	PlainText
		x:9
		y:17
		value:"Robot Job"
	DropDown
		x:1
		y:19
		1:"       Offensive Robot    "
		2:"       Defensive Robot    "
		3:"Offensive and Defensive Robot"
		4:"       Did Nothing     "

TeleScreen2
	name: Tele2
	PlainText
		x:0
		y:5
		value:"Pass to Human?"
	Checkbox
		x:15
		y:4
		name:humanPass
	PlainText
		x:0
		y:7
		value:"Goal Defense?"
	Checkbox
		x:15
		y:6
		name:goalDefense
	PlainText
		x:0
		y:9
		value:"Fouls"
	Counter
		x:6
		y:9
		increment:x
		name:fouls
	PlainText
		x:12
		y:9
		value:"Tech Fouls"
	Counter
		x:23
		y:9
		increment:y
	PlainText
		x:0
		y:12
		value:"Driver Skill"
	Rating
		x:12
		y:12
		name:driverRating
	PlainText
		x:0
		y:14
		value:"Defense"
	Rating
		x:12
		y:14
		name:defenseRating
	PlainText
		x:0
		y:16
		value:"Pickup"
	Rating
		x:12
		y:16
		name:pickupRating
	Finish
		x:24
		y:20


Scouting/DS Stuff/ScoutTestFiles/config.2011.txt
Configuration
	InputFile: DS.txt # The input file of match numbers and teams playing.
	Foo: file.out # The file to put the output of the scouting.

Screens
	1:HomeScreen
	2:AutoScreen
	3:TeleScreen
	4:TeleScreen2

HomeScreen
	name: Home
	ShowScoutingInitial

AutoScreen
	name: Auto
	PlainText
		x:0
		y:5
		value:"Starting Place"
	PlainText
		x:0
		y:6
		value:"Left"
	PlainText
		x:6
		y:6
		value:"Middle"
	PlainText
		x:14
		y:6
		value:"Right"
	Radio
		x1:1
		y1:7
		x2:8
		y2:7
		x3:15
		y3:7
		name:startPos
	PlainText
		x: 0
		y: 10
		value: "Penalties"
	UpDown
		x:10
		y:10
		name:autopenalties
	PlainText
		x: 0
		y: 13
		value: "Overall Rating"
	Rating
		x: 14
		y: 13
		name: autoRating
	PlainText
		x: 0
		y: 17
		value: "Uber-Tube on:"
	PlainText
		x: 14
		y: 17
		value: "Top"
	PlainText
		x: 18
		y: 17
		value: "Mid"
	PlainText
		x: 22
		y: 17
		value: "Low"
	PlainText
		x: 26
		y: 17
		value: "None"
	Radio
		x1:14
		y1:18
		x2:18
		y2:18
		x3:22
		y3:18
		x4:27
		y4:18
		name:uberPlacement
	PlainText
		x:0
		y:20
		value:"Did nothing/disabled"
	Checkbox
		x:11
		y:21
		name:autoNothing

TeleScreen
	name: Tele
	PlainText
		x:1
		y:4
		value:"Robot job:"
	DropDown
		x: 1
		y: 5
		1: "    Offensive Robot    "
		2: "    Defensive Robot    "
		3: "Offensive AND Defensive"
		4: "      Did Nothing      "
		name:positions
	PlainText
		x:23
		y:7
		value:"Penalties"
	UpDown
		x:27
		y:9
		name:telepenalties
	PlainText
		x:0
		y:7
		value: "Scored on:"
	PlainText
		x:0
		y:8
		value:"Top"
	PlainText
		x:6
		y:8
		value:"Mid"
	PlainText
		x:12
		y:8
		value:"Low"
	PlainText
		x:18
		y:8
		value:"???"
	Counter
		x: 0
		y: 9
		increment:x
		name: scorestop
	Counter
		x: 6
		y: 9
		increment:y
		name: scoresmid
	Counter
		x: 12
		y: 9
		increment:b
		name: scoreslow
	Counter
		x:18
		y:9
		name: unknownscores
		increment:a
	PlainText
		x:0
		y:11
		value:"Completes logos?"
	Checkbox
		x:5
		y:12
		name:logos
	Checkbox
		x:25
		y:12
		name:groundPickup
	PlainText
		x:0
		y:14
		value:"Misses"
	Counter
		x:7
		y:14
		name:misses
		increment:r
	PlainText
		x:0
		y:17
		value:"Grabs:"
	PlainText
		x:6
		y:16
		value:"Circle"
	PlainText
		x:14
		y:16
		value:"Square"
	PlainText
		x:22
		y:16
		value:"Triangle"
	Checkbox
		x:8
		y:17
		name:grabsCirc
	Checkbox
		x:16
		y:17
		name:grabsSqr
	Checkbox
		x:25
		y:17
		name:grabsTri
	PlainText
		x:0
		y:19
		value:"Maneuverability"
	PlainText
		x:8
		y:20
		value:"Rating"
	Rating
		x:14
		y:20
		name:maneuverRating
	PlainText
		x:0
		y:22
		value:"Manipulability"
	PlainText
		x:0
		y:23
		value:"(of arm)"
	Rating
		x:14
		y:22
		name:armRating
	PlainText
		x:18
		y:11
		value:"Picks up from"
	PlainText
		x:18
		y:12
		value:"Ground"
	PlainText
		x:12
		y:13
		value:"Status"
	DropDown
		x:12
		y:14
		1:"  Active  "
		2:"Red-carded"
		3:" Disabled "
		name:status

TeleScreen2
	name: Tele2
	PlainText
		x:3
		y:5
		value:"Minibot Success rating:"
	DropDown
		x:1
		y:6
		1:"   Does not have Minibot   "
		2:"       No Deployment       "
		3:"    Did not make to top    "
		4:" Made to top AFTER TIME UP "
		5:"  Made to top within time  "
		6:"Other Bots Deployed Instead"
		name:minibotRating
	DropDown
		x:1
		y:17
		1:"1st place"
		2:"2nd place"
		3:"3rd place"
		4:"4th place"
		5:"no finish"
		name:miniPlacement
	PlainText
		x:0
		y:8
		value:"Deployment rating"
	PlainText
		x:0
		y:11
		value:"Aproximate secs to climb"
	Rating
		x:0
		y:9
		name:deploySpeed
	Timer
		x:25
		y:11
		press:a
		name:minibotSpeed
	PlainText
		x:0
		y:14
		value:"Any minibot penalties"
	Checkbox
		x:22
		y:14
		name:any_penalties
	PlainText
		x:1
		y:16
		value:"Rank"
	Finish
		x:24
		y:20


Scouting/DS Stuff/DSCode/source/config.c
/* Copyright, Nicholas Carlini. All rights reserved. */

/** Read values from config file.
  *
  * A @c Config describes screens and questions to display on a Nintendo DS.
  * A @c Config contains a sequence of sections. Sections contain either nested sections or key/value pairs.
  *
  * USAGE:
  * Call @c configure to read a file and return the @c Config* it represents.
  * Call @c config_lookup_section to lookup the section for a name.
  * Call @c config_lookup_key to lookup the value for a name.
  * 
  * SYNTAmX:
  *
  * EXAMPLE:
  * For example:
  *     Section1
  *         Section 1A
  *             key1: value1
  *             key2: value2
  *         Section 1B
  *             key3: value1
  *     Section2
  *         Section 2A
  *             key4: value1
  *         Section 2B
  *             key5: value1
  *             key6: value2
  *    ...
  *
**/

#include <stdio.h>
#include <stdlib.h>
#include <nds.h>

#include "config.h"
#include "parse.h"
#include "lex.h"


char*  result = "\
Configuration\n\
	InputFile: DS.txt # The input file of match numbers and teams playing.\n\
	Foo: file.out # The file to put the output of the scouting.\n\
\n\
Screens\n\
	1:InitialScreen\n\
	2:AutoScreen\n\
	3:TeleScreen\n\
	4:EndScreen\n\
\n\
InitialScreen\n\
	name: Init\n\
	ShowScoutingInitial\n\
\n\
AutoScreen\n\
	name: Auto\n\
	PlainText\n\
		x:0\n\
		y:5\n\
		value:\"Starting Place\"\n\
	PlainText\n\
		x:0\n\
		y:6\n\
		value:\"Left\"\n\
	PlainText\n\
		x:6\n\
		y:6\n\
		value:\"Middle\"\n\
	PlainText\n\
		x:14\n\
		y:6\n\
		value:\"Right\"\n\
	Radio\n\
		x1:1\n\
		y1:7\n\
		x2:8\n\
		y2:7\n\
		x3:15\n\
		y3:7\n\
		name:startPos\n\
	PlainText\n\
		x: 0\n\
		y: 10\n\
		value: \"Penalties\"\n\
	UpDown\n\
		x:10\n\
		y:10\n\
		name:autopenalties\n\
	PlainText\n\
		x: 0\n\
		y: 13\n\
		value: \"Overall Rating\"\n\
	Rating\n\
		x: 14\n\
		y: 13\n\
		name: autoRating\n\
		width: 1\n\
	PlainText\n\
		x: 0\n\
		y: 17\n\
		value: \"Uber-Tube on:\"\n\
	PlainText\n\
		x: 14\n\
		y: 17\n\
		value: \"Top\"\n\
	PlainText\n\
		x: 18\n\
		y: 17\n\
		value: \"Mid\"\n\
	PlainText\n\
		x: 22\n\
		y: 17\n\
		value: \"Low\"\n\
	PlainText\n\
		x: 26\n\
		y: 17\n\
		value: \"None\"\n\
	Radio\n\
		x1:14\n\
		y1:18\n\
		x2:18\n\
		y2:18\n\
		x3:22\n\
		y3:18\n\
		x4:27\n\
		y4:18\n\
		name:uberPlacement\n\
	PlainText\n\
		x:0\n\
		y:20\n\
		value:\"Did nothing/disabled\"\n\
	Checkbox\n\
		x:11\n\
		y:21\n\
		name:autoNothing\n\
\n\
TeleScreen\n\
	name: Tele\n\
	PlainText\n\
		x:1\n\
		y:4\n\
		value:\"Robot job:\"\n\
	DropDown\n\
		x: 1\n\
		y: 5\n\
		1: \"    Offensive Robot    \"\n\
		2: \"    Defensive Robot    \"\n\
		3: \"Offensive AND Defensive\"\n\
		4: \"      Did Nothing      \"\n\
		name:positions\n\
	PlainText\n\
		x:23\n\
		y:7\n\
		value:\"Penalties\"\n\
	UpDown\n\
		x:27\n\
		y:9\n\
		name:telepenalties\n\
	PlainText\n\
		x:0\n\
		y:7\n\
		value: \"Scored on:\"\n\
	PlainText\n\
		x:0\n\
		y:8\n\
		value:\"Top\"\n\
	PlainText\n\
		x:6\n\
		y:8\n\
		value:\"Mid\"\n\
	PlainText\n\
		x:12\n\
		y:8\n\
		value:\"Low\"\n\
	PlainText\n\
		x:18\n\
		y:8\n\
		value:\"???\"\n\
	Counter\n\
		x: 0\n\
		y: 9\n\
		increment:x\n\
		name: scorestop\n\
	Counter\n\
		x: 6\n\
		y: 9\n\
		increment:y\n\
		name: scoresmid\n\
	Counter\n\
		x: 12\n\
		y: 9\n\
		increment:b\n\
		name: scoreslow\n\
	Counter\n\
		x:18\n\
		y:9\n\
		name: unknownscores\n\
		increment:a\n\
	PlainText\n\
		x:0\n\
		y:11\n\
		value:\"Completes logos?\"\n\
	Checkbox\n\
		x:5\n\
		y:12\n\
		name:logos\n\
	Checkbox\n\
		x:25\n\
		y:12\n\
		name:groundPickup\n\
	PlainText\n\
		x:0\n\
		y:14\n\
		value:\"Misses\"\n\
	Counter\n\
		x:7\n\
		y:14\n\
		name:misses\n\
		increment:r\n\
	PlainText\n\
		x:0\n\
		y:17\n\
		value:\"Grabs:\"\n\
	PlainText\n\
		x:6\n\
		y:16\n\
		value:\"Circle\"\n\
	PlainText\n\
		x:14\n\
		y:16\n\
		value:\"Square\"\n\
	PlainText\n\
		x:22\n\
		y:16\n\
		value:\"Triangle\"\n\
	Checkbox\n\
		x:8\n\
		y:17\n\
		name:grabsCirc\n\
	Checkbox\n\
		x:16\n\
		y:17\n\
		name:grabsSqr\n\
	Checkbox\n\
		x:25\n\
		y:17\n\
		name:grabsTri\n\
	PlainText\n\
		x:0\n\
		y:19\n\
		value:\"Maneuverability\"\n\
	PlainText\n\
		x:8\n\
		y:20\n\
		value:\"Rating\"\n\
	Rating\n\
		x:14\n\
		y:20\n\
		name:manueverRating\n\
	PlainText\n\
		x:0\n\
		y:22\n\
		value:\"Manipulability\"\n\
	PlainText\n\
		x:0\n\
		y:23\n\
		value:\"(of arm)\"\n\
	Rating\n\
		x:14\n\
		y:22\n\
		name:amRating\n\
	PlainText\n\
		x:18\n\
		y:11\n\
		value:\"Picks up from\"\n\
	PlainText\n\
		x:18\n\
		y:12\n\
		value:\"Ground\"\n\
	PlainText\n\
		x:12\n\
		y:13\n\
		value:\"Status\"\n\
	DropDown\n\
		x:12\n\
		y:14\n\
		1:\"  Active  \"\n\
		2:\"Red-carded\"\n\
		3:\" Disabled \"\n\
		name:status\n\
\n\
EndScreen\n\
	name: End\n\
	PlainText\n\
		x:3\n\
		y:5\n\
		value:\"Minibot Success rating:\"\n\
	DropDown\n\
		x:1\n\
		y:6\n\
		1:\"   Does not have Minibot   \"\n\
		2:\"       No Deployment       \"\n\
		3:\"    Did not make to top    \"\n\
		4:\" Made to top AFTER TIME UP \"\n\
		5:\"  Made to top within time  \"\n\
		6:\"Other Bots Deployed Instead\"\n\
		name:minibotRating\n\
	DropDown\n\
		x:1\n\
		y:17\n\
		1:\"1st place\"\n\
		2:\"2nd place\"\n\
		3:\"3rd place\"\n\
		4:\"4th place\"\n\
		5:\"no finish\"\n\
		name:miniPlacement\n\
	PlainText\n\
		x:0\n\
		y:8\n\
		value:\"Deployment rating\"\n\
	PlainText\n\
		x:0\n\
		y:11\n\
		value:\"Aproximate secs to climb\"\n\
	Rating\n\
		x:0\n\
		y:9\n\
		name:deploySpeed\n\
	Timer\n\
		x:25\n\
		y:11\n\
		press:a\n\
		name:minibotSpeed\n\
	PlainText\n\
		x:0\n\
		y:14\n\
		value:\"Any minibot penalties\"\n\
	Checkbox\n\
		x:22\n\
		y:14\n\
		name:any_penalties\n\
	PlainText\n\
		x:1\n\
		y:16\n\
		value:\"Rank\"\n\
	Finish\n\
		x:24\n\
		y:20\n\
";




/**
 * Load the file into a character array.
 * Fill in the length into the len pointer.
 */
void loadFromFile(char** dest, int* len) {
  FILE* fp = fopen("config.txt", "r");
	
	if (fp == NULL) {
		*dest = result;
		*len = strlen(result);
		return;
	}

  /* Get the file size. */
  fseek(fp, 0L, SEEK_END);
  int size = ftell(fp);
  *len = size;
  fseek(fp, 0L, SEEK_SET);

  *dest = (char*)malloc(sizeof(char)*(size+1));
  char* buf = *dest;
  int chr = 0;
  int index = 0;
  /* Read a character. Make sure that we don't smash the heap. */
  while ((chr = fgetc(fp)) != EOF && index+1 < size) buf[index++] = chr;
  buf[index] = 0;

  fclose(fp);
}

/**
 * Print the configuration file out recursively.
 * Used for debugging.
 */
void printConf(Config* conf, int indent) {
  if (conf == NULL) return;
  printf("%d: %s\n", indent, conf->key);  
  KVPair* kv = conf->values;
  while (kv != NULL) {
    printf("[kv] %s: %s\n", kv->key, kv->value);
    kv = kv->next;
  }
  printConf(conf->children, indent+1);
  while (1) {
    conf = conf->next;
    if (conf == NULL) break;
    printConf(conf, indent);
  }
}

/**
 * Lookup the value for the given key in the linked list of key/value pairs.
 */
char* config_lookup_key(Config* conf, char* key) {
  KVPair* values = conf->values;
  while (values != NULL) {
		//printf("Comparing %s %s\n", key, values->key);
    if (strcmp(key, values->key) == 0) {
      return values->value;
    }
    values = values->next;
  }
  return NULL;
}

/**
 * Lookup the section for the given name in the linked list of children.
 */
Config* config_lookup_section(Config* conf, char* name) {
  Config* child = conf->children;
  while (child != NULL) {
    //printf("&%s/%s/%d*", name, child->key, strcmp(name, child->key));
    if (strcmp(name, child->key) == 0) {
      return child;
    }
    child = child->next;
  }
  return NULL;
}

/**
 * Do all the configuration. 
 * Load everything from the file, lex and parse the input, and then return the result.
 *
 * At the end of this function, there have been several calls to malloc:
 *  (1) The string representing the file has been allocated.
 *  (2) Each Token has been allocated.
 *  (3) Each Config struct and KV Pair has been allocated.
 * 	(a) Inside each config node, the name is a newly allocated string.
 *	(b) As well, in each KV Pair the key and value are allocated.
 *
 * These must be freed by the end of the program.
 */
Config* configure() {
  char* put;
  int len;
  loadFromFile(&put, &len);
  Token* tokenList = lex(put, len);
	
  Config* configFile = parse(tokenList, put);
  return configFile;
}
